docs/spec_draft.md

```diff
@@ -102,26 +102,192 @@ never produced as outputs.
+
+Within the grid, the processes are mostly independent from each other - they
+have separate operation statuses, separate input/intermediate/output values and
+most of the ops are executed separately between processes, with the exception of
+a small number of collective ops described below.
+
+Given that execution of most of the ops is only using values from the same
+process, it is usually unambiguous to refer to these values by their names.
+However, when describing semantics of collective ops, that is insufficient, and
+we use the notation `name@process_id` to refer to the value `name` within a
+particular process. (From that perspective, unqualified `name` can be viewed as
+a shorthand for `name@(replica_id(), partition_id())`).
+
+The execution order across processes is implementation-defined, except for the
+synchronization introduced by collective ops as described below.
+
+### Collective ops
+
+There are five collective ops in StableHLO: `all_gather`, `all_reduce`,
+`all_to_all`, `collective_permute` and `reduce_scatter`. All these ops split
+the processes in the StableHLO grid into **StableHLO process groups** and
+execute a joint computation within each process group, independently from other
+process groups.
+
+Within each process group, collective ops may introduce a synchronization
+barrier. Further formalization, e.g. elaborating on when exactly this
+synchronization happens, how exactly the processes arrive at this barrier,
+and what happens if they don't, is TBD.
+
+If the process group involves cross-partition communication, i.e. there are
+processes in the process group whose partition ids are different, then execution
+of the collective op needs a **StableHLO channel**, and the collective op must
+provide a positive `channel_id` of type `si64`. Further formalization, e.g.
+where these channel ids are coming from and how they are synchronized between
+programs, is TBD. Cross-replica communication doesn't need channels.
+
+The computations performed by the collective ops are specific to individual ops
+and are described in individual op sections below. However, the strategies by
+which the grid is split into process groups are shared between these ops and are
+described in this section. More formally, StableHLO supports the following
+four strategies.
+
+1) **cross_replica** (as in "only cross-replica communications will be happening
+within each process group"). This strategy takes `replica_groups` - a list of
+lists of replica ids - and computes a Cartesian product of `replica_groups` by
+`partition_ids`. `replica_groups` must have unique elements and cover all
+`replica_ids`. More formally:
+
+```Python
+def cross_replica(replica_groups: List[List[ReplicaId]]) -> List[List[ProcessId]]:
+  for replica_group in replica_groups:
+    for partition_id in partition_ids:
+      process_group = []
+      for replica_id in replica_group:
+        process_group.append((replica_id, partition_id))
+      yield process_group
```

Comment:
I am sorry, I did not understand the significance of `yield` here.

Reply:
I used it as a way to cut down on boilerplate - instead of defining `process_groups = []` at the top of the function, appending to it here and then returning it at the bottom of the function, we can just yield.